src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,183 @@
+    return _committedState->getTC();
+}
+
+void
+SDCPUThread::issueInstruction(StaticInstPtr inst, TheISA::PCState cur_pc)
+{
+    const InstSeqNum seq_num = lastCommittedInstNum + inflightInsts.size();
+
+    shared_ptr<InflightInst> dynamic_inst =
+        make_shared<InflightInst>(getThreadContext(), seq_num, inst, cur_pc);
+
+    dynamic_inst->populateDataDependencies(inflightInsts, lastUses);
+    // TODO update lastUses to include the new instruction's uses
+    inflightInsts.push_back(dynamic_inst);
+    // TODO request execution for any instructions that are ready.
+}
+
+void
+SDCPUThread::onBeginFetch()
+{
+    // Retrieve the value of the speculative PC as a raw Addr value, aligned
+    // to MachInst size, but offset by the amount of data we've fetched thus
+    // far, since multiple fetches may be needed to feed one decode.
+    const Addr pc_addr = _speculativePC.instAddr();
+    const Addr req_vaddr = (pc_addr & BaseCPU::PCMask) + fetchOffset;
+
+    const RequestPtr fetch_req = std::make_shared<Request>();
+
+    // Set Ids for tracing requests through the system
+    fetch_req->taskId(_cpuPtr->taskId());
+    fetch_req->setContext(_committedState->contextId()); // May need to store
+                                                        // speculative variant
+                                                        // if we allow out-of-
+                                                        // order execution
+                                                        // between context
+                                                        // changes?
+
+    fetch_req->setVirt(0, req_vaddr, sizeof(TheISA::MachInst),
+                       Request::INST_FETCH, _cpuPtr->instMasterID(), pc_addr);
+
+    if (!_cpuPtr->requestInstAddrTranslation(fetch_req,
+        [this](Fault f, const RequestPtr& r) {onPCTranslated(f, r);})) {
+        panic("The CPU rejected my fetch translate request and I haven't been "
+            "programmed to know how to continue!!!");
+        // TODO reschedule a fetch for the future?
+    }
+}
+
+void
+SDCPUThread::onPCTranslated(Fault fault, const RequestPtr& req)
+{
+    if (fault != NoFault) {
+        // TODO handle fault (Store fault for commit stage, then stop further
+        // fetching)
+    }
+
+    // No changes to the request before asking the CPU to handle it if there is
+    // not a fault. Let the CPU generate the packet.
+    if (!_cpuPtr->requestInstruction(req,
+        [this](Fault f, TheISA::MachInst i) {
+            onInstructionFetched(f, i);
+        })) {
+        panic("The CPU rejected my fetch request and I haven't been programmed"
+            " to know how to continue!!!");
+    }
+}
+
+void
+SDCPUThread::onInstructionFetched(Fault fault, const TheISA::MachInst inst)
+{
+    if (fault != NoFault) {
+        // TODO handle fault
+    }
+
+    // Capure the PC state at the point of instruction retrieval.
+    TheISA::PCState pc_state = _speculativePC;
```

Comment:
Umm... Not sure if this is right. Couldn't this be some other PC if this fetch took a long time?

Reply:
This should be correct until we want to implement out-of-order/speculative pre-fetching.